compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -259,6 +260,44 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
     }
 }
 
+pub fn remove_duplicate_unreachable_blocks<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    struct OptApplier<'tcx> {
+        tcx: TyCtxt<'tcx>,
+        duplicates: FxIndexSet<BasicBlock>,
+    }
+
+    impl<'tcx> MutVisitor<'tcx> for OptApplier<'tcx> {
+        fn tcx(&self) -> TyCtxt<'tcx> {
+            self.tcx
+        }
+
+        fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+            for target in terminator.successors_mut() {
+                if self.duplicates.contains(target) {
+                    *target = self.duplicates[0];
+                }
+            }
+
+            self.super_terminator(terminator, location);
+        }
+    }
+
+    let unreachable_blocks = body
+        .basic_blocks
+        .iter_enumerated()
+        .filter_map(|(block, bb)| {
+            // CfgSimplifier::simplify leaves behind some unreachable basic blocks without a
+            // terminator. Those blocks will be deleted by remove_dead_blocks, but we run just
+            // before then so we need to handle missing terminators.
+            if bb.terminator.is_some() && bb.is_empty_unreachable() { Some(block) } else { None }
+        })
```

Comment:
This is a nit-pick, but can this be turned into `filter`+`map`? I find code with `filter_map` and an `bool::then_some`-ish code harder to read than just `filter` and `map` by themselves.

```suggestion
        .filter(|(_, bb)| {
            // CfgSimplifier::simplify leaves behind some unreachable basic blocks without a
            // terminator. Those blocks will be deleted by remove_dead_blocks, but we run just
            // before then so we need to handle missing terminators.
            bb.terminator.is_some() && bb.is_empty_unreachable()
        })
        .map(|(block, _)| block)
```